apollo-router/src/query_planner/bridge_query_planner.rs

```diff
@@ -215,14 +215,23 @@ impl BridgeQueryPlanner {
         let selections = self.parse_selections(key.0.clone()).await?;
 
         if selections.contains_introspection() {
-            // If we have only one operation containing a single root field `__typename`
-            if selections.contains_only_typename() {
+            // If we have only one operation containing only the root field `__typename`
+            // (possibly aliased or repeated). (This does mean we fail to properly support
+            // {"query": "query A {__typename} query B{somethingElse}", "operationName":"A"}.)
+            let output_keys_for_typenames = if selections.operations.len() == 1 {
+                selections.operations[0].is_only_typenames_with_output_keys()
+            } else {
+                None
+            };
```

Comment:
Something like this would be better

Reply:
@bnjjj Er wait. This change is semantically different if the doc has multiple operations. Choosing to evaluate the first operation even if a different one was selected via operationName seems wrong. My code also is wrong in that case but it's wrong by potentially failing to do the typename directly or maybe getting an empty plan error, but an incorrect null or error seems better than an incorrect successful execution of the wrong op, no?

(Perhaps I should just fix it to be right instead of wrong, idk. Actually this whole codepath lets you execute invalid documents, ignores skip/include, etc... perhaps the query planner just needs a new "static" node type that returns some JSON without a fetch. But I still think actively running the wrong operation is worse then failing ...)

Reply:
Or alternatively given that format_response appears to handle some of this logic maybe the fix is just to delete the special case here and be ok with an empty query plan?